### PR TITLE
Add bintrees to travis and as optional dependency

### PR DESCRIPTION
**Pull request by MSeifert04**
It's given as ["optional dependency"](https://github.com/astropy/astropy/blob/master/astropy/table/bst.py#L670) in the code but isn't listed.

### CODE REVIEW

**Comment by bsipocz** [comment]
Would it make any sense to also raise a warning when someone uses these engines but doesn't have bintrees installed?

**Comment by pllim** [comment]
I'll let someone else set the milestone. I still get confused on doc milestones.

**Comment by bsipocz** [comment]
@pllim - Most docs PR can go into the bugfix (unless the docs is about something that's new feature in master). Or at least I see these as fixes to the current docs.

**Comment by MSeifert04** [comment]
> Would it make any sense to also raise a warning when someone uses these engines but doesn't have bintrees installed?

@bsipocz I agree but the module code just defaults to the Python implementation without any warning. So I guess that either was an oversight or intended behavior given that the index for table is still listed as [provisional](http://docs.astropy.org/en/stable/table/indexing.html).

**Comment by bsipocz** [comment]
> @bsipocz I agree but the module code just defaults to the Python implementation without any warning. So I guess that either was a major oversight or intended behavior, given that the index for table is still listed as provisional.

Yes, I seeing them defaulting to the pure python implementation made me wonder about this. I feel at that point it'd be good to give a heads up to the user that they are not actually using what they thought they did.

**Comment by MSeifert04** [comment]
@bsipocz Maybe better to open an issue about the warning if bintrees is not installed. I don't want to fix that in this PR (I'm a bit out of my depth if it comes to `Table` internals) but it could be a good enhancement.

**Comment by MSeifert04** [comment]
Should I mention that it's needed for _really_ fast `FastRBT` or `FastBST` index engine instead of the "possible speedups with table indexing"?

**Comment by bsipocz** [comment]
I would leave that as is, as based on this notebook they're not always faster: http://nbviewer.jupyter.org/github/mdmueller/astropy-notebooks/blob/master/table/indexing-profiling.ipynb

**Comment by MSeifert04** [comment]
Yeah, I didn't mean overall speedup (I just remembered why I made it so imprecise...). I meant that it will speedup *only* the `FastBST` and `FastRBT` because these would otherwise default to the slower pure-Python `BST`.

**Comment by taldcroft** [comment]
It is certainly a concern that the bintrees-based code is not tested in CI.  I guess the first question is how easy it would be to include this in the optional-dependency CI testing.

If it is not tested then it would be useful in this PR to update the docs to reflect that (e.g. a red warning box).  

On my long-term to-do list is allowing an indexed Table to be fully serialized to ECSV, FITS, HDF5.  Even though I was the instigator and mentor for this GSoC project (adding indexing) I have not found much use for it because of the lack of persistence.  From that perspective having only the SortedArray option would be better because that is something that can be serialized in a straightforward way.  Which is a roundabout way of floating the idea of just deprecating then removing the binary-tree based options.

**Comment by MSeifert04** [comment]
I added `bintrees` to a travis build. Let's see how that works out...

**Comment by MSeifert04** [comment]
I'm not sure if I understand `pytest.fixture` with `params` correctly does that do the same as parametrize? Did I change it "correctly"?

**Comment by MSeifert04** [comment]
The travis failure is because I've cancelled the OSX build the change build succeeded. If the change does what I hope it does that means `bintrees`-index is tested now?

**Comment by MSeifert04** [comment]
@taldcroft Given that it passes the tests I don't think we need "red warning boxes". But could you review the PR?

**Comment by MSeifert04** [comment]
Anything I can do here?

**Comment by pllim** [comment]
Let's ping @taldcroft again for an approval.

File: astropy/table/tests/test_index.py [hunk 1]
```diff
@@ -4,18 +4,30 @@
...
 from ...extern.six.moves import range
 
-
-@pytest.fixture(params=[BST, FastRBT, SortedArray])
-def engine(request):
-    return request.param
+try:
+    import bintrees
+except ImportError:
+    _HAS_BINTREES = False
```
Comment (taldcroft) [inline astropy/table/tests/test_index.py]: Not a big deal but typically (at least from my experience) these `HAS_` vars are not prefixed with `_`.

File: astropy/table/tests/test_index.py [hunk 2]
```diff
@@ -4,18 +4,30 @@
...
+else:
+    _HAS_BINTREES = True
+
+
+if _HAS_BINTREES:
```
Comment (taldcroft) [inline astropy/table/tests/test_index.py]: Not sure if there is some pytest magic happening, but could you just define `engines` as a list based on `HAS_BINTREES` and then have `@pytest.fixture(params=engines)`.  I.e. don't repeat the whole function definition twice.  Just wondering.

**Comment by MSeifert04** [comment]
@taldcroft Thanks for the feedback. The suggestions make sense and I changed it accordingly. 👍

**Comment by bsipocz** [comment]
As it was approved already, I'm merging this now. Thanks @MSeifert04!

**Pull request merged**